MarcusBarnes/mik
# Issue 156

**Github issue**:  https://github.com/MarcusBarnes/mik/issues/156

# What does this Pull Request do?

This PR offers an initial implementation of shutdown hooks, a feature described in #156.

# What's new?
Adds a .ini array, [WRITER]['shutdownhooks'] that should contain commands to be executed just before mik exits. After the postwritehooks loop completes, the elements of this array will be invoked, one after the other.

## Possible problems
As with post-write hooks, the shutdown hooks loop begins without checking to be sure the post-write processes have completed, therefore, anything that relies on the output of the post-write stage will need to do it's own completion-checking (perhaps a log or other artifact left by post-write to signify completion).

# How should this be tested?

Add the following to your .ini. When mik exits the config file should have a current last-modified time.

    [WRITER]
    shutdownhooks[] = "touch"

For a less-trivial test, try the included xslt runner.

    [WRITER]
    shutdownhooks[] = "/usr/bin/php extras/scripts/shutdownhooks/apply_xslt_with_saxon.php"

which will require some additional configuration:

    [XSLT]
    working_dir   = 'hnoc-delcroix/hnoc-xslt'
    step_thru     = TRUE
    stylesheets[] = 'extras/lsu/xsl/titleNonSort.xsl'
    stylesheets[] = 'extras/lsu/xsl/subjectSplit.xsl'
    stylesheets[] = 'extras/lsu/xsl/dateCreatedSplit.xsl'
    stylesheets[] = 'extras/lsu/xsl/locationMerge.xsl'
    stylesheets[] = 'extras/lsu/xsl/blankNodes.xsl'
    stylesheets[] = 'extras/lsu/xsl/OrderedTemplates.xsl'

# Additional Notes:

##Documentation snippet:

### Shutdown hooks

Similar to post-write hooks, but as the name implies, shutdown hooks run just before mik exits.
Shutdown hooks also differ in that they are run sequentially. That is, the first shutdown hook must
exit before the second is invoked.

Register shutdown hooks in .ini in the `[WRITER]` section:

	 shutdownhooks[] = "/usr/bin/php extras/scripts/shutdownhooks/shutitdown.php"


# Interested parties
@mjordan @MarcusBarnes
